### DISCUSSION TITLE
Reducing use of SOQL in synchronous context

### DISCUSSION THREAD

**Discussion by LawrenceLoz** [jongpie/NebulaLogger#303] [Ideas]
Hi there,

Firstly big kudos to author/authors and contributors on this project - a really rich and intuitive set of features and some great configuration flexibility.

Running a few experiments and looking at debug logs, I've noticed that when the Logger object class is called several queries are run within the transaction which contribute to the limit of 100 SOQL statements - in my tests this is 4 queries and I think would be 5 for logs being made for community users. Whilst this isn't an especially high, in an ideal world given how widely a logging framework should be used it would be great to reduce the utilisation and improve synchronous performance where possible.

There are a couple of areas resulting in queries being run, and I wonder whether it's viable to refactor to avoid these:

1. Line 2556 in Logger initialises LogEntryEventHandler: `LogEntryEventHandler().executeBeforeInsert(logEntryEvents);` As part of this, TAG_ASSIGNMENT_RULES variable is set from the results of a CMDT query, which includes flitering on SObjectField__r.QualifiedApiName (the bit which makes the CMDT query contribute to limits like a normal SOQL object query). This variable isn't used until after PEs are created so feels like this should be called only at that stage
2. There are methods queryOrganization(), queryUser(), queryAuthSession() and queryNetwork() in the LogEntryEventBuilder, which all query for details to populate fields on the event. Could it make sense instead to populate just the user id on the event and avoid checking for this extra information until LogEntryEventHandler is processing the events? The downside of this approach is that events would be missing the extra info so an external system consuming these wouldn't have the detail, so maybe the choice of whether these queries are made before or after events should be configurable in settings

I'm happy to create issues and help out with PRs suggesting changes for some or all of these areas if you agree there's a better approach? Or let me know if I'm missing something about how any of these can be avoided today?

Also a heads up I'm hoping to touch on Nebula Logger as part of a talk I'm co-presenting at London's Calling soon (https://www.londonscalling.net/sessions/session-apex-frameworks-for-architects/). Let me know if you'd like more insight into how this would be included (or if for any reason you'd prefer Nebula Logger not to be included!)

Again, thanks for the great work - I've developed a similarly architected logging framework myself and familiar with many of the challenges you've addressed really well here!

**Comment by jongpie** [comment]
Hi @LawrenceLoz, this is some incredible feedback, thanks for taking the time to share your findings & thoughts! Currently, the code runs 3-5 queries synchronously (`queryAuthSession()` and `queryNetwork()` may not run for all users/scenarios). Originally, it was [just 1 query for the current user](https://github.com/jongpie/NebulaLogger/blob/v0.13.0-beta/src/classes/Logger.cls#L9), but as new features have been added over time, the number of queries has grown. I've thought about some possible ways to mitigate it, but have yet to actually implement any solutions for it (besides caching the queries within each transaction). I fully agree that although 3-5 queries isn't especially high, it'd still be nice to reduce that count where possible.

I think there are a couple of changes that could be combined to improve the query usage:

- I like your idea of providing a configurable way to control if the queries run synchronously or asynchronously. By default, I'd want to keep running the queries synchronously, as there are some features that could be negatively impacted if the queries only run asynchronously, but for orgs that don't leverage every feature, being able to control when the queries run would be a great option. I would probably want to control this org-wide with a new record in the custom metadata type `LoggerParameter__mdt`. Another option would be to instead control it with the custom hierarchy setting `LoggerSettings__c` so it can be configured per profile/user, but it would probably be a bit more complex to handle it per user, instead of having just a single org-wide option.
- For orgs that want to continue running the queries synchronously, I think the query count could still be optimized by using platform cache. I had [explored using platform cache](https://github.com/jongpie/NebulaLogger/tree/feature/platform-cache/nebula-logger/main/cache) about a year ago for [a different use case](https://github.com/jongpie/NebulaLogger/issues/18) that was abandoned, but some of the platform caching code/configuration could be re-used. With platform cache, I think that the results for `queryOrganization()`, `queryUser()`, `queryNetwork()`, and the constant `LogEntryEventHandler.TAG_ASSIGNMENT_RULES` could all be cached in an org cache, and the results of `queryAuthSession()` could be cached in a session cache. Occasionally, the queries would have to re-run (if the results aren't in the cache or the cache has expired), but this would help eliminate running the queries when the results have been cached, while still populating the platform event fields synchronously.
- The `LogEntryEventHandler.TAG_ASSIGNMENT_RULES` could be optimized by lazy-loading with a custom getter (something I already do for several other constants/variables). That would avoid/delay running the query until it's actually needed. Something like this:
   ```java
   private static List<LogEntryTagRule__mdt> TAG_ASSIGNMENT_RULES {
        get {
            if (TAG_ASSIGNMENT_RULES == null) {
                TAG_ASSIGNMENT_RULES = getTagAssignmentRules();
            }
            return TAG_ASSIGNMENT_RULES;
        }
        set;
    }
   ```

Let me know your thoughts on these possible changes!

And that's awesome to hear that you might discuss Nebula Logger at London's Calling! I'd absolutely love to hear more about what you'll be covering - feel free to [message me on LinkedIn](https://www.linkedin.com/in/jongpie/) if you'd like to discuss it outside of GitHub.

Thanks again for your feedback!

**Reply by LawrenceLoz** [reply]
Hi @jongpie, thanks for the considered feedback and sharing a bit of background on the queries!

In relation to your suggestions for approach:

- I can understand your preference to run queries before firing an event by default, an agree setting whether these happen synchronously or not would only warrant an org-wide setting so LoggerParameter__mdt sounds like the right place. I wonder if it's worth separating the configuration options so that there's one flag for each group of data? This probably isn't much more complexity but gives some flexibility which might be helpful for some.
- Completely with you that platform cache is well suited to improving query performance and the approach sounds good. One thing to be conscious of is I think platform cache is only provided free to Enterprise Edition orgs and above, and I don't know what would happen trying to deploy a partition to an org which wasn't licensed for it. I can't imagine there are many Professional or Essentials customers using Nebula directly, but not sure what would happen if it's used within a package? (or if that's actually a concern!?)
- Lazy loading TAG_ASSIGNMENT_RULES is as great idea and addresses this nicely

Let me know if you'd like me to raise issues for any of these changes? It might not be for a couple of weeks but also happy to help out in the dev too if it's useful.

And I'll drop you a message about London's Calling separately with a bit more info.

Thanks and all the best
Lawrence

**Reply by jongpie** [reply]
Hi @LawrenceLoz, just a few follow up thoughts on your last response:

- I like the idea of having separate configuration for each query, and I agree that it would not greatly increase the complexity. There will be a new `LoggerParameter__mdt` record per query method, something like:
  - `queryAuthSession()` uses an instance of `LoggerParameter__mdt` called "QueryAuthSessionSynchronously" (or something similar), with a default value of "true"
  - `queryOrganization()` uses an instance of `LoggerParameter__mdt` called "QueryOrganizationSynchronously (or something similar), with a default value of "true"
  - (repeat the same convention for each query method)
- That's a fair point about platform cache allocations for Professional & Essentials. As far as I know, there aren't any Professional or Essentials orgs using Nebula Logger, but it could be that I'm just not aware of those orgs. I'll do some additional research for platform cache before moving forward with creating any issues or implementing any changes.
- Great, I'll include lazy-loading of `TAG_ASSIGNMENT_RULES` in a future release, should be fairly easy to make that change

As far as raising issues & implementing these changes, I'm happy to handle writing up the issue for running the queries async & developing the necessary changes. Some of these query changes will align with some other optimizations that I'm planning to implement over the next few weeks (related to mocking records to improve test speeds), so this should be fairly straightforward to include in an upcoming release.

Looking forward to hearing more about the presentation at London's Calling!

**Reply by LawrenceLoz** [reply]
Hi @jongpie, agree with all your points and great to hear the query changes should fit with some of the other enhancements you have planned. Happy to help out reviewing when you're ready if it's useful.

I also ran the debug logs of the transaction through Apex Log Analyzer (https://marketplace.visualstudio.com/items?itemName=financialforce.lana) and thought you might be interested in the output:

![image](https://user-images.githubusercontent.com/32834730/167012910-b404c7ec-2d39-4a2e-a8e2-e48290489329.png)

Looks like there should be a good performance gain during logs publishing from avoiding the tag entry rule query, so will definitely be nice to see that running outside of the sync context.

I noted a few other things on the screenshot which thought were interesting. Depending on your appetite for performance improvements, there might be some value in making other small changes. I think there's a possibility to reduce CPU overhead (very slightly!) by caching results of UserInfo.getUserSession() - this is run during Logger.getTransactionId() and again from LogEntryEventBuilder.queryAuthSession() taking 4ms and then 3ms. There would be a saving too in allowing orgs to disable capturing the stack trace - this took 6ms during the Logger.debug() call and time would presumably increase linearly with each call - it's a nice feature but orgs bumping against CPU limits might value being able to disable.

Other than this the profile looks very reasonable, I couldn't see anything else which looks avoidable or especially intensive

**Reply by jongpie** [reply]
Hi @LawrenceLoz - thanks for doing this analysis, this is very useful info to have, and it's analysis that I have not done in a while, so it's fantastic seeing these numbers. I'd like to try to incorporate all of your suggestions, including caching `UserInfo.getSessionId()` and being able to disable the `parseStackTrace()` method. I've created issue #311 to capture all of the changes - I still need to clean up/expand upon some details, but I think I've got all of your suggestions mentioned within that issue. 

I'm planning to start incorporating some of these changes ASAP - I'll definitely keep you updated on my progress (some of the changes may be done over multiple releases), and I would love to get your thoughts if you're interested in doing a code review of the changes.
